prototype_source/maskedtensor_adagrad.rst

```diff
@@ -0,0 +1,182 @@
+and when it is just used to compress zeros.
+We'll also compare and contrast this with equivalent code written using MaskedTensor.
+In the end the code snippets are repeat without additional comments to show the difference in brevity.
+
+Set up
+------
+
+.. code-block:: python
+
+    import torch
+    from torch.masked.maskedtensor import masked_tensor
+
+    # Some hyperparameters
+    eps = 1e-10
+    clr = 0.1
+
+    i = torch.tensor([[0, 1, 1], [2, 0, 2]])
+    v = torch.tensor([3, 4, 5], dtype=torch.float32)
+    grad = torch.sparse_coo_tensor(i, v, [2, 4])
+
+Original sparse implementation
+------------------------------
+
+First, let's break down the current implementation of 
+`Adagrad (functional) <https://github.com/pytorch/pytorch/blob/6c2f235d368b697072699e5ca9485fd97d0b9bcc/torch/optim/_functional.py#L16-L51>`__
+in PyTorch currently:
+
+.. code-block:: python
+    :linenos:
+
+    def _make_sparse(grad, grad_indices, values):
+        size = grad.size()
+        if grad_indices.numel() == 0 or values.numel() == 0:
+            return torch.empty_like(grad)
+        return torch.sparse_coo_tensor(grad_indices, values, size)
+
+    # We don't support sparse gradients
+    param = torch.arange(8).reshape(2, 4).float()
+    state_sum = torch.full_like(param, 0.5)  # initial value for state sum
+
+    grad = grad.coalesce()  # the update is non-linear so indices must be unique
+    grad_indices = grad._indices()
+    grad_values = grad._values()
+    # pow(2) has the same semantics for both sparse and dense memory layouts since 0^2 is zero
+    state_sum.add_(_make_sparse(grad, grad_indices, grad_values.pow(2)))
+
+    # We take care to make std sparse, even though state_sum clearly is not.
+    # This means that we're only applying the gradient to parts of the state_sum
+    # for which it is specified. This even drives the point home a lot more that
+    # the passed gradient is not sparse, but masked.
+    std = state_sum.sparse_mask(grad)
+
+    # We currently dodge all these concerns using the private method values.
+    std_values = std._values().sqrt_().add_(eps)
+
+    # We currently don't support div for sparse Tensors because zero / zero is
+    # not well defined. For a MaskedTensor undefined / undefined is undefined.
+    param.add_(_make_sparse(grad, grad_indices, grad_values / std_values), alpha=-clr)
+
+Line 24 is where we have a very important divergence.
+
+The addition of eps should technically be applied to all values but instead is only applied to specified values.
+Here we're using sparsity as a semantic extension and to enforce a certain pattern of defined and undefined values.
+If parts of the values of the gradient are zero, they are still included if materialized even though they
+could be compressed by other sparse storage layouts.
+This is technically quite brittle even though someone could argue that eps is always very small.
```

Comment:
```suggestion
This is technically quite brittle! That said, someone could argue that eps is always very small, so it might not matter so much in practice.
```

Would recommend splitting this into two sentences to get your point across first + then allow the competing comment later. I wasn't sure how eps being very small counteracted the point, and it'd be cool if you could offer an explanation.